dhall-nix/src/Dhall/Nix.hs

```diff
@@ -738,3 +728,129 @@ dhallToNix e =
     loop (Embed x) = absurd x
+
+-- | Previously we turned @<Foo | Bar>.Foo@ into @{ Foo, Bar }: Foo@,
+-- but this would not work with <Frob/Baz>.Frob/Baz (cause the slash is not a valid symbol char in nix)
+-- so we generate @union: union."Frob/Baz"@ instead.
+--
+-- If passArgument is @Just@, pass the argument to the union selector.
+unionChoice :: Text -> Maybe NExpr -> NExpr
+unionChoice chosenKey passArgument =
+  let selector = Fix (Nix.NSelect Nothing (Nix.mkSym "u") (mkDoubleQuotedIfNecessary chosenKey :| []))
+  in Nix.Param "u" ==>
+     case passArgument of
+       Nothing -> selector
+       Just arg -> selector @@ arg
+
+
+-- | Double-quote a field name (record or union). This makes sure it’s recognized as a valid name by nix, e.g. in
+--
+--  @{ "foo/bar" = 42; }."foo/bar" }@
+--
+-- where
+--
+-- @{ foo/bar = 42; }.foo/bar@ is not syntactically valid nix.
+--
+-- This is only done if necessary (where “necessary” is not super defined right now).
+mkDoubleQuotedIfNecessary :: Text -> NKeyName r
+mkDoubleQuotedIfNecessary key =
+    if Text.all simpleChar key
+    then StaticKey (VarName key)
+    else DynamicKey (Plain (DoubleQuoted [Plain key]))
+    where
+        simpleChar c = isAsciiLower c || isAsciiUpper c
+
+
+-- | Nix does not support symbols like @foo/bar@, but they are allowed in dhall.
+-- So if they happen, we need to encode them with an ASCII escaping scheme.
+--
+-- This is copied/inspired by the Z-Encoding scheme from GHC, see
+-- https://hackage.haskell.org/package/zenc-0.1.2/docs/Text-Encoding-Z.html
+--
+-- Original Source is BSD-3-Clause, Copyright (c)2011, Jason Dagit
+zEncodeSymbol :: Text -> Text
+zEncodeSymbol = zEncodeString
+
+-- | The basic encoding scheme is this:
+
+--   * Alphabetic characters (upper and lower) and digits
+--         all translate to themselves;
+--         except 'Z', which translates to 'ZZ'
+--         and    'z', which translates to 'zz'
+--
+--   * Most other printable characters translate to 'zx' or 'Zx' for some
+--         alphabetic character x
+--
+--   * The others translate as 'znnnU' where 'nnn' is the decimal number
+--         of the character
+--
+-- @
+--         Before          After
+--         --------------------------
+--         Trak            Trak
+--         foo-wib         foozmwib
+--         \>               zg
+--         \>1              zg1
+--         foo\#            foozh
+--         foo\#\#           foozhzh
+--         foo\#\#1          foozhzh1
+--         fooZ            fooZZ
+--         :+              ZCzp
+-- @
+zEncodeString :: Text -> Text
+zEncodeString cs = case Text.uncons cs of
+    Nothing -> Text.empty
+    Just (c, cs') ->
+        encodeDigitChar c
+        <> Text.concatMap encodeChar cs'
+
+-- | Whether the given characters needs to be z-encoded.
+needsEncoding :: Char -> Bool
+needsEncoding 'Z' = True
+needsEncoding 'z' = True
+needsEncoding c   = not
+                  ( isAsciiLower c
+                  || isAsciiUpper c
```

Comment:
Perhaps reuse the `simpleChar` function here

Reply:
Hm, I’d have to think about whether they are semantically the same.

Reply:
Indeed, thinking about it they are not the same, since we either double quote where necessary (using the `simpleChar` filter), or z-encode where necessary (internal symbols that cannot be quoted). In particular, z-encoded values should never land inside a struct for example.

I think this made me notice a place where I’d forgotten the encoding, namely `Project`.

Reply:
What I mean is that right now `simpleChar c` is defined as `isAsciiLower c || isAsciiUpper c`, so it seems like you should be able to replace `isAsciiLower c || isAsciiUpper c` right here with `simpleChar c`

Reply:
I still feel like you can reuse the `simpleChar` function here:

```haskell
needsEncoding c = not (simpleChar c)
```